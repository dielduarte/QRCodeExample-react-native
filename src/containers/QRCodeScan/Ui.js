import React, { Fragment } from 'react';

import Button from '../../components/Button';
import Camera from '../../components/Camera';

const Ui = ({ showCamera, setShowCamera, onBarCodeRead }) => {
  return (
    <Fragment>
      <Button title="Scan" onPress={() => setShowCamera(!showCamera)} />
      {showCamera && <Camera onBarCodeRead={onBarCodeRead} />}
    </Fragment>
  );
};

export default Ui;
