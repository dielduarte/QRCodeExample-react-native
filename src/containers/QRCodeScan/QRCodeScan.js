/**
 * Isolate the bussines logic and Ui
 * I chosed recompose todo that in a easy way
 * in a real scenario, we can do it for example to:
 * 1 - use the bussines logic of the component on the web/mobile apps
 * without need to rewrite it.
 * 2 - keep the code more clean and testable.
 */

import { withHandlers, withState, compose } from 'recompose';
import Ui from './Ui';

export const onBarCodeRead = props => data => {
  const { setShowCamera, showCamera } = props;

  alert(data.data);
  setShowCamera(!showCamera);
};

const QRCodeScan = compose(
  withState('showCamera', 'setShowCamera', false),
  withHandlers({ onBarCodeRead })
);

export default QRCodeScan(Ui);
