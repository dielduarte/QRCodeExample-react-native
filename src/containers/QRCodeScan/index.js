import QRCodeScan from './QRCodeScan';
export default QRCodeScan;
