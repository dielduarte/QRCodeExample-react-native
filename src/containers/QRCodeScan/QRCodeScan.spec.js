import React from 'react';
import { shallow } from 'enzyme';
import QRCodeScan, { onBarCodeRead } from './QRCodeScan';

describe('QRCodeScan', () => {
  it('should render correctly', () => {
    const component = shallow(<QRCodeScan />).dive();
    expect(component).toMatchSnapshot();
  });

  describe('onBarCodeRead', () => {
    it('should called an alert with the data value and setShowCamera prop passing false as value', () => {
      const props = {
        setShowCamera: jest.fn(),
        showCamera: true
      };
      const data = {
        data: 'https://lisk.io'
      };

      onBarCodeRead(props)(data);

      expect(props.setShowCamera).toHaveBeenCalledWith(false);
      expect(global.alert).toHaveBeenCalledWith(data.data);
    });
  });
});
