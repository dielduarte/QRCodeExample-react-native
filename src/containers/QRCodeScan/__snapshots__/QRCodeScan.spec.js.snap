// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeScan should render correctly 1`] = `
<Ui
  onBarCodeRead={[Function]}
  setShowCamera={[Function]}
  showCamera={false}
/>
`;
